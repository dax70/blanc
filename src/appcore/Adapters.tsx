import * as React from 'react';
import { Menu as AppMenu, 
    MenuItem as AppMenuItem, 
    MenuItemKind as AppMenuItemKind
} from './Menus';
import { 
    Menu, MenuItem, MenuDivider,
    Popover, Position, Icon 
} from '@blueprintjs/core';

export function convertItem(item: AppMenuItem, index: number) {
    switch (item.kind) {
        case AppMenuItemKind.separator:
            return <MenuDivider key={index}/>;
        case AppMenuItemKind.text:
        default:
            const { text, iconName } = item; 
            return (<MenuItem key={index} text={text} iconName={iconName} />);
    }
}

export interface MenuAdapterProps {
    menu: AppMenu;
}

export class MenuAdapter extends React.Component<MenuAdapterProps, {}> {
    constructor(props: MenuAdapterProps) {
        super(props);
    }

    render() {
        // Convert to Menu system
        const { menu } = this.props;
        const items = menu.getItems().map(convertItem);
        return (
            <Popover 
                className="main-menu"
                position={Position.BOTTOM}
                inheritDarkTheme={false}
                content={
                    <Menu className="pt-elevation-1">
                        {items}
                    </Menu>
                } 
                target={                    
                    <button className="pt-button pt-minimal pt-icon-document">
                        Menu
                        <Icon 
                            className="pt-align-right" 
                            iconName="caret-down" 
                        />                                
                    </button>
                } 
            />
        );
    }
}