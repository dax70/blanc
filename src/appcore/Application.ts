import { Menu } from '../appui/Menu';
import { Item } from './Item';

class Application {
  mainMenu: Menu;

  constructor() {
    // Something here
  }

  load() {
    // Load code
    // notify all components
  }

  getSelection(): Item | null {
    // get selected item
    return null;
  }

  onSelectionChanged() {
    // notify selection changed
  }

  unload() {
    // Unload code
    // notify all components
  }
}

export default Application;