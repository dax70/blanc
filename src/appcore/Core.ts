import { IconName as BPIconName } from '@blueprintjs/core/dist/generated/iconName';

type ButtonEvent = {

};

// reference to the Blueprint Icons
export type IconName = BPIconName;

export interface ButtonProps {
    text: string;
    click?: (e: ButtonEvent) => void;
    iconName?: IconName;
}

export class Button {
    text: string;
    click?: (e: ButtonEvent) => void;
    iconName?: string;

    constructor(props: ButtonProps) {
        Object.assign(this, props);
    }
}