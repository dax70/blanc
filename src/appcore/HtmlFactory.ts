import {
 DocumentNode, 
 HtmlNode, 
 HtmlNodeProps, 
 HtmlTag, 
 HtmlNodeKind, 
 NodeProps, 
 ChildNodes 
} from './document';

export class HTMLElement implements DocumentNode, HtmlNode {
  kind: HtmlNodeKind = 'HTMLElement';  
  tag: HtmlTag;
  props?: NodeProps;
  children?: ChildNodes;

  constructor(nodeProps: HtmlNodeProps) {
    const { tag, props, children } = nodeProps;
    this.tag = tag;
    this.props = props;
    this.children = children;
  }
   
  clone() {
    return Object.assign({}, this);
  }
}

export default class Html {

  static create(tag: HtmlTag, props?: NodeProps, children?: ChildNodes) {
    return new HTMLElement({ tag, props, children });
  }
}