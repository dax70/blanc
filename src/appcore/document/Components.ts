// Must either define all props as optinal
// or leave as any
export type ComponentProps = {
};

export type ChildrenComponent = {} | Array<object> | string | null | undefined;

export enum ComponentKind {
  Component = 'Component',
  Html = 'HTMLElement',
  Text = 'Text'
}

export interface DocumentComponent {
  kind: ComponentKind;
  isSelected?: boolean;
  props?: ComponentProps;
  children?: ChildrenComponent;
}

// More?
export type HtmlTag = 'a' | 'div' | 'p' | 'span';

export type HtmlComponentProps = {
  tag: HtmlTag;  
  props?: ComponentProps;
  children?: ChildrenComponent;
};

export interface HtmlComponent extends DocumentComponent, HtmlComponentProps {
  kind: ComponentKind.Html;
}