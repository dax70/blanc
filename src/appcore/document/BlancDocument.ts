import { DocumentComponent } from './Components';
import { EventHandler, IndexedArgs, Subject, Subscription } from '../Subscriptions';

export type DocumentCallBack = EventHandler<DocumentComponent>;

export type Selection = {
  index: number;
  node: DocumentComponent;
};

export type DocumentIndexedCallBack = EventHandler<IndexedArgs<DocumentComponent>>;

export default class BlancDocument {
    nodes: Array<DocumentComponent>;
    selection?: Selection;
    addSubject: Subject<DocumentComponent>;
    insertSubject: Subject<IndexedArgs<DocumentComponent>>;
    removeSubject: Subject<IndexedArgs<DocumentComponent>>;
     
    constructor() {
      this.nodes = [];

      // TODO consider moving to key based structure
      // and lazy initialize subjects - aggregate subscriptions.
      this.addSubject = new Subject();
      this.insertSubject = new Subject();
      this.removeSubject = new Subject();      
    } 

    addComponent(node: DocumentComponent) {
      this.nodes.push(node);
      this.addSubject.next(node);
    }

    insertComponent(index: number, node: DocumentComponent) {
      this.nodes.splice(index, 0, node);
      this.insertSubject.next({index, item: node});
    }

    removeComponent(node: DocumentComponent) {
      const index  = this.nodes.indexOf(node);

      if (index < 0) {
        throw new Error('Unable to find component to remove.');
      }

      this.nodes.splice(index, 1);
      this.removeSubject.next({index, item: node});
    }

    setSelection(selection: Selection) {
      const prevSelection = this.selection;
      if (prevSelection) {
        const { node: oldNode } = prevSelection;
        if (oldNode) {
          oldNode.isSelected = false;
        }
      }

      const { node } = selection;
      node.isSelected = true;
      this.selection = selection;
    }

    getSelection() {
      return this.selection;
    }

    //#region Events

    onDidAddComponent(handler: DocumentCallBack): Subscription {
      return this.addSubject.subscribe(handler);
    }

    onDidInsertComponent(handler: DocumentIndexedCallBack): Subscription {
      return this.insertSubject.subscribe(handler);
    }

    onDidRemoveComponent(handler: DocumentIndexedCallBack): Subscription {
      return this.removeSubject.subscribe(handler);
    }

    //#endregion

    getItems() {
      return this.nodes;
    }
 
    serialize() {
      // Serialize
    }

    deserialize(json: string) {
      // Deserialize
    }
}