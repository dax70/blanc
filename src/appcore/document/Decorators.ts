import { DocumentComponent } from './Components';

export enum CardDepth {
  zero,
  one,
  two,
  three,
  four
}

export interface Card extends DocumentComponent { 
  depth: CardDepth;
}