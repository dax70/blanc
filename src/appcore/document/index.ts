import {
  ComponentKind,
  DocumentComponent, 
  HtmlComponent, 
  HtmlComponentProps, 
  HtmlTag, 
  ComponentProps, 
  ChildrenComponent 
 } from './Components';

import BlancDocument from './BlancDocument';

export {
  BlancDocument,
  ComponentKind,
  DocumentComponent, 
  HtmlComponent, 
  HtmlComponentProps, 
  HtmlTag, 
  ComponentProps, 
  ChildrenComponent 
 };