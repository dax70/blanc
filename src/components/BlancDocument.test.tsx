import * as React from 'react';
import * as renderer from 'react-test-renderer'; 

import BlancDocument from './BlankDocument';
import { BlancDocument as DocModel } from '../appcore/document';
import HtmlFactory from '../appcore/HtmlFactory';

test('BlancDocument renders div Hello', () => {

  const docContent = new DocModel();
  const divContent = HtmlFactory.create(
    'div',
    {},
    'Hello'
  );

  docContent.addComponent(divContent);

  const blancDoc = renderer.create( 
    <BlancDocument content={docContent} />
  ).toJSON();

  expect(blancDoc).toMatchSnapshot();
});

test('BlancDocument mock doc part', () => {
  
    const docContent = new DocModel();

    const divContent = HtmlFactory.create(
      'div',
      {},
      'Hellow from inside content'
    );

    const mockEl = {
      kind: 'HTMLElement', 
      tag: 'div',
      props: { style: { color: 'blue' }}, 
      children: 'Hello from object'
    };

    const root = HtmlFactory.create(
      'div',
      {},
      ['Hello text', divContent, mockEl ]
    );

    docContent.addComponent(root);
  
    const blancDoc = renderer.create(
      <BlancDocument content={docContent} />
    ).toJSON();

    expect(blancDoc).toMatchSnapshot();
  });