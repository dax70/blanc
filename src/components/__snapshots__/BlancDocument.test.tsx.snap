// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlancDocument mock doc part 1`] = `
<div
  className="bl-doc"
>
  <div>
    <span>
      Hello text
    </span>
    <div>
      Hellow from inside content
    </div>
    <div
      style={
        Object {
          "color": "blue",
        }
      }
    >
      Hello from object
    </div>
  </div>
</div>
`;
